system/lib/wasmfs/backends/ignore_case_backend.cpp

```diff
@@ -0,0 +1,126 @@
+// were accesible. Then, we would store the original case there but search
+// ignoring case.
+//
+class IgnoreCaseDirectory : public MemoryDirectory {
+  using BaseClass = MemoryDirectory;
+  std::shared_ptr<Directory> baseDirectory;
+
+public:
+  IgnoreCaseDirectory(std::shared_ptr<Directory> base, backend_t proxyBackend)
+    : BaseClass(base->locked().getMode(), proxyBackend), baseDirectory(base) {}
+
+  std::shared_ptr<File> getChild(const std::string& name) override {
+    return BaseClass::getChild(normalize(name));
+  }
+
+  std::shared_ptr<DataFile> insertDataFile(const std::string& name,
+                                           mode_t mode) override {
+    auto name2 = normalize(name);
+    auto child = baseDirectory->insertDataFile(name2, mode);
+    insertChild(name2, child);
+    return child;
+  }
+
+  std::shared_ptr<Directory> insertDirectory(const std::string& name,
+                                             mode_t mode) override {
+    auto name2 = normalize(name);
+    auto baseChild = baseDirectory->insertDirectory(name2, mode);
+    auto child = std::make_shared<IgnoreCaseDirectory>(baseChild, getBackend());
+    insertChild(name2, std::static_pointer_cast<File>(child));
+    return child;
+  }
+
+  std::shared_ptr<Symlink> insertSymlink(const std::string& name,
+                                         const std::string& target) override {
+    auto name2 = normalize(name);
+    auto child = baseDirectory->insertSymlink(name2, target);
+    insertChild(name2, child);
+    return child;
+  }
+
+  int insertMove(const std::string& name, std::shared_ptr<File> file) override {
+    return BaseClass::insertMove(normalize(name), file);
+  }
+
+  bool removeChild(const std::string& name) override {
+    auto name2 = normalize(name);
+    BaseClass::removeChild(name2);
+    return baseDirectory->removeChild(name2);
+  }
+
+  size_t getNumEntries() override { return BaseClass::getNumEntries(); }
+
+  // TODO: preserve original name, denormalize, and use it here
+  std::vector<Directory::Entry> getEntries() override {
+    return BaseClass::getEntries();
+  }
+
+  // TODO: preserve original name, denormalize, and use it here
+  std::string getName(std::shared_ptr<File> file) override {
+    return BaseClass::getName(file);
+  }
+
+  bool maintainsFileIdentity() override { return true; }
+};
+
+class IgnoreCaseBackend : public Backend {
+  backend_t backend;
+
+public:
+  IgnoreCaseBackend(std::function<backend_t()> createBackend) {
+    backend = createBackend();
+  }
+
+  std::shared_ptr<DataFile> createFile(mode_t mode) override {
+    return backend->createFile(mode);
+  }
+
+  std::shared_ptr<Directory> createDirectory(mode_t mode) override {
+    return std::make_shared<IgnoreCaseDirectory>(backend->createDirectory(mode),
+                                                 this);
+  }
+
+  std::shared_ptr<Symlink> createSymlink(std::string target) override {
```

Comment:
There is no way to create a symlink directly from a backend. Only through `Directory::insertSymlink()`.

Reply:
The `create*` methods on backends are mostly for creating new backend roots to mount onto other backends. We might want to consider removing `createFile` and only supporting `createDirectory` so that only directories can be backend roots. That's more similar to how native file systems work. On the other hand, if we do want to keep `createFile`, it probably makes sense to add `createSymlink` as well.

Reply:
Ok, I see, so `createSymlink` is just for parity with `createFile`.